internal/pkg/zeromq/client.go

```diff
@@ -0,0 +1,258 @@
+	var errorsSubscribe []error
+	var err error
+
+	for idx, topic := range topics {
+		client.subscribers[idx], err = client.subscribeTopic(&topic)
+		if err != nil {
+			errorsSubscribe = append(errorsSubscribe, err)
+		}
+	}
+
+	if len(errorsSubscribe) == 0 {
+		return nil
+	}
+
+	var errorStr string
+	for _, err := range errorsSubscribe {
+		errorStr = errorStr + fmt.Sprintf("%s  ", err.Error())
+	}
+	return errors.New(errorStr)
+}
+
+func (client *zeromqClient) subscribeTopic(topic *messaging.TopicChannel) (*zeromqSubscriber, error) {
+
+	subscriber := zeromqSubscriber{}
+	msgQueueURL := client.getSubscribeMessageQueueURL()
+	if err := subscriber.init(msgQueueURL, topic); err != nil {
+		return nil, err
+	}
+
+	go func(topic *messaging.TopicChannel) {
+
+		_ = subscriber.connection.SetSubscribe(topic.Topic)
+		for {
+
+			payloadMsg, err := subscriber.connection.RecvMessage(0)
+
+			select {
+			   case <- client.quitSubscribe:
+				   _ = subscriber.connection.SetLinger(time.Duration(0))
+				   _ = subscriber.connection.Close()
+				   return
+			   default:
+			}
+
+			// This may occur if try to receive before publisher has established the socket. Just try again.
+			if err != nil && err.Error() != "resource temporarily unavailable" {
+				continue
+			}
+
+			var payload []byte
+			switch msgLen := len(payloadMsg); msgLen {
+			case 1:
+				payload = []byte(payloadMsg[singleMessagePayloadIndex])
+			case 2:
+				payload = []byte(payloadMsg[multiMessagePayloadIndex])
+			default:
+				client.messageErrors <- fmt.Errorf("found more than 2 incoming messages (1 is no topic, 2 is topic and message), but found: %d", msgLen)
+				continue
+			}
+
+			msgEnvelope := messaging.MessageEnvelope{}
+			err = json.Unmarshal(payload, &msgEnvelope)
+			if err != nil {
+				client.messageErrors <- err
+				continue
+			}
+
+			topic.Messages <- &msgEnvelope
+		}
+	}(&subscriber.topic)
+
+	return &subscriber, nil
+}
+
+func (client *zeromqClient) Disconnect() error {
+
+	var disconnectErrs []error
+
+	if client.quitSubscribe != nil {
+		close(client.quitSubscribe)
+	}
+
+	for _, subscriber := range client.subscribers {
+		// This forces the receive calls to unblock
+		err := subscriber.context.Term()
+
+		if err != nil {
+			disconnectErrs = append(disconnectErrs, err)
+		}
+	}
+
+	if client.publisher != nil {
+		_ = client.publisher.SetLinger(time.Duration(0))
+		err := client.publisher.Close()
+		if err != nil {
+			disconnectErrs = append(disconnectErrs, err)
+		}
+	}
+
+	if client.messageErrors != nil {
+		close(client.messageErrors)
+	}
+
+	for _, subscriber := range client.subscribers {
+		if subscriber.topic.Messages != nil {
+			close(subscriber.topic.Messages)
+		}
+	}
+
+	if len(disconnectErrs) == 0 {
+		return nil
+	}
+
+	var errorStr string
+	for _, err := range disconnectErrs {
+		if err != nil {
+			errorStr = errorStr + fmt.Sprintf("%s  ", err.Error())
+		}
+	}
+
+	return errors.New(errorStr)
+}
+
```

Comment:
In the functions below, I think some of this assembly should be put into the HostInfo type itself.

Example:
https://github.com/edgexfoundry/edgex-go/blob/66d0c5babe8ad42eb0b4effdc94eb9c4fcee08e2/internal/pkg/config/types.go#L57

Reply:
Good idea.  Also did some refactoring to pull the types from the interface.go into a types namespace, that allowed us to consolidate the factory interface into the messaging namespace.